tokio/src/process/unix/pidfd_reaper.rs

```diff
@@ -0,0 +1,310 @@
+use crate::{
+    io::{interest::Interest, PollEvented},
+    process::{
+        imp::{orphan::Wait, OrphanQueue},
+        kill::Kill,
+    },
+    util::error::RUNTIME_SHUTTING_DOWN_ERROR,
+};
+
+use libc::{syscall, SYS_pidfd_open, __errno_location, ENOSYS, PIDFD_NONBLOCK};
+use mio::{event::Source, unix::SourceFd};
+use std::{
+    fs::File,
+    future::Future,
+    io,
+    marker::Unpin,
+    ops::Deref,
+    os::unix::io::{AsRawFd, FromRawFd, RawFd},
+    pin::Pin,
+    process::ExitStatus,
+    sync::atomic::{AtomicBool, Ordering::Relaxed},
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+struct Pidfd {
+    fd: File,
+}
+
+impl Pidfd {
+    fn open(pid: u32) -> Option<Pidfd> {
+        // Store false (0) to reduce executable size
+        static NO_PIDFD_SUPPORT: AtomicBool = AtomicBool::new(false);
+
+        if NO_PIDFD_SUPPORT.load(Relaxed) {
+            return None;
+        }
+
+        unsafe {
+            let fd = syscall(SYS_pidfd_open, pid, PIDFD_NONBLOCK);
+            if fd == -1 {
+                let errno = *__errno_location();
+
+                if errno == ENOSYS {
+                    NO_PIDFD_SUPPORT.store(true, Relaxed)
+                }
+
+                None
+            } else {
+                Some(Pidfd {
+                    fd: File::from_raw_fd(fd as i32),
+                })
+            }
+        }
```

Comment:
Generally, it is preferred to have a separate unsafe block for each unsafe operation, and to annotate each block with a `// SAFETY:` comment justifying its correctness.

Reply:
Thanks!

I've reduced the `unsafe {}` block to each function calls and also replaced use of `__errno_location()` with `io::Error::last_os_error().raw_os_error().unwrap()`